src/Repository.php

```diff
@@ -0,0 +1,200 @@
+	 */
+	public function selectFirst(array $condition)
+	{
+		$data = $this->dba->selectFirst(static::$table_name, [], $condition);
+
+		if (!$data) {
+			throw new HTTPException\NotFoundException(static::class . ' record not found.');
+		}
+
+		return $this->create($data);
+	}
+
+	/**
+	 * Populates a Collection according to the condition.
+	 *
+	 * Chainable.
+	 *
+	 * @param array $condition
+	 * @param array $params
+	 * @return Collection
+	 * @throws \Exception
+	 */
+	public function select(array $condition = [], array $params = [])
+	{
+		$models = $this->selectModels($condition, $params);
+
+		return new static::$collection_class($models);
+	}
+
+	/**
+	 * Populates the collection according to the condition. Retrieves a limited subset of models depending on the boundaries
+	 * and the limit. The total count of rows matching the condition is stored in the collection.
+	 *
+	 * Chainable.
+	 *
+	 * @param array $condition
+	 * @param array $params
+	 * @param int?  $max_id
+	 * @param int?  $since_id
+	 * @param int   $limit
+	 * @return Collection
+	 * @throws \Exception
+	 */
+	public function selectByBoundaries(array $condition = [], array $params = [], int $max_id = null, int $since_id = null, int $limit = self::LIMIT)
+	{
+		$condition = DBA::collapseCondition($condition);
+
+		$boundCondition = $condition;
+
+		if (isset($max_id)) {
+			$boundCondition[0] .= " AND `id` < ?";
+			$boundCondition[] = $max_id;
+		}
+
+		if (isset($since_id)) {
+			$boundCondition[0] .= " AND `id` > ?";
+			$boundCondition[] = $since_id;
+		}
+
+		$params['limit'] = $limit;
+
+		$models = $this->selectModels($boundCondition, $params);
+
+		$totalCount = DBA::count(static::$table_name, $condition);
+
+		return new static::$collection_class($models, $totalCount);
+	}
+
+	/**
+	 * This method updates the database row from the model.
+	 *
+	 * @param Model $model
+	 * @return bool
+	 * @throws \Exception
+	 */
+	public function save(Model $model)
+	{
+		return $this->dba->update(static::$table_name, $model->toArray(), ['id' => $model->id], true);
+	}
+
+	/**
+	 * This method creates a new database row and returns a model if it was successful.
+	 *
+	 * @param Model $model
+	 * @return Model|bool
+	 * @throws \Exception
+	 */
+	public function insert(array $fields)
+	{
+		$return = $this->dba->insert(static::$table_name, $fields);
+
+		if ($return) {
+			$fields['id'] = $this->dba->lastInsertId();
+			$return = $this->create($fields);
+		}
+
+		return $return;
+	}
+
+	/**
+	 * Deletes the model record from the database.
+	 *
+	 * @param Model $model
+	 * @return bool
+	 * @throws \Exception
+	 */
+	public function delete(Model &$model)
+	{
+		if ($success = $this->dba->delete(static::$table_name, ['id' => $model->id])) {
+			$model = null;
+		}
+
+		return $success;
+	}
+
+	/**
+	 * Base instantiation method, can be overriden to add specific dependencies
+	 *
+	 * @param array $data
+	 * @return Model
+	 */
+	protected function create(array $data)
```

Comment:
What about "really" creating a new model instance, which isn't saved yet?

My approach was:
```php
// creates a new instance with possible data **except** a database id
$Intro = DI::intro()->create($data);
// or
$Intro = DI::intro()->create();
$Intro->uid = 2;

// save it
// checks that an insert instead of an update is needed
// sets the lastInsertId of the insert as is of the current model
DI::intro()->save($Intro);

Reply:
The current `save()` method expects the model's `id` to be set, that's why I created the `insert()` not to have to check for its existence.

Reply:
We could simply this to save, so it is up to the Repository what's to do :) and the caller has just to call save.

Reply:
I don't think it would be simpler. It wold be more convenient because we would have only one method to call, but the method itself would be more complex than both current methods combined. So the convenience would have a cost.

Reply:
why is there a cost?
```php
if (empty($model->id)) {
   $this->dba->insert(...)
   $model->id = $this->dba->lastInsertId();
} else {
   $this->dba->update(...);
}
```

It's a more "OOP" way, because an object is saved, nevermind if it's in a DB or inside a textfile or inside a cache. It's up to the repository what to do with it

Reply:
Everytime you have a `if` or a `switch`, there's an extra complexity cost because of the branching code paths. If you have two distint methods each doing their own thing, it is less complex and more explicit.

Sometimes you can't avert it, but when you can avoid a `if` that isn't a data check I think you should avoid it.

Reply:
Note that the `insert` will allow for easier transition from `DBA::insert` and can still be used the OOP way:
```
$intro = DI::intro()->insert();
$intro->uid = 2;
DI::intro()->save($intro);
```
Although it isn't great for multiple reasons. I see what you mean about the more OOP way but I'd really like to keep the complexity down.

Reply:
Hm .. so it's up to the caller if he creates and inserts or select and updates.. This could maybe lead to a problem if we will extend collections with newly created objects.

But because I do want to make progress (todos based on this PR ;-), we'll see how it works. If necessary, this is a point which we can re-structure later if "really" needed :)

Reply:
Got it, I'll try to button it up tonight. Expect a few API changes.